### DISCUSSION TITLE
pre-commit script with `interactive: true` unable to capture user input from console/tty

### DISCUSSION THREAD

**Discussion by marques-work** [evilmartians/lefthook#709] [Q&A] [answered]
I have a simple script that I want to run as a pre-commit hook. It checks if the user is on the `main` branch, and asks the user to confirm if they really meant to commit directly to `main` (not everyone on my team has a git prompt that shows the current branch so I thought this would be an easy guard). I don't want to outright deny commits to `main` because there are valid use cases, so I just want to verify with the user.

Anyway, I've set `interactive: true` and even tried that in combination with `use_stdin: true` but the script just hangs and doesn't seem to be receiving any input from the keyboard. I thought this would be simple and straightforward, but apparently not.

#### I'm using lefthook `v1.6.10`

My repo is an internal github repo, so I can't link it, but here is what the script looks like:

```bash
#!/bin/bash
# vim: sts=2 sw=2 et ai

set -euo pipefail

exec </dev/tty

main() {
  # Get the current branch name
  local branch="$(git rev-parse --abbrev-ref HEAD)"

  # Check if the current branch is main
  if [ 'main' = "$branch" ]; then
    read -r -p 'You are on the `main` branch. Are you sure you want to commit here? [y/N]: ' answer

    if [ 'y' != "$(tr '[:upper:]' '[:lower:]' <<< "$answer")" ]; then
      printf '\e[31;1m[ERROR] %s\e[0m\n' 'Aborted commit!'
      exit 1
    fi
  fi
}

main
```
And yes, I have tried variations without `exec < /dev/tty` or with an explicit `read -r answer < /dev/tty` and neither works.

My `lefthook.yml` looks like:

```yaml
pre-commit:
  scripts:
    check-branch.sh:
      runner: bash
      interactive: true
```

@mrexox any ideas?

**Comment by mrexox** [comment]
@marques-work , this config should work:

```yml
pre-commit:
  scripts:
    check-branch.sh:
      runner: bash
      use_stdin: true
      interactive: true
```

And no tweak with /dev/tty is required in the script itself. For example, I could make it work with the following script:

```bash
read -r -p '[y/N]' answer

if [ "$answer" = "y" ]; then
  echo 'YES'
  exit 0
else
  echo 'NO'
  exit 1
fi
```

I think it is kind of confusing that both `interactive` and `use_stdin` are required. It feels like if we have `interactive` we probably want to use STDIN too 🤔 . I will rethink this behavior. 

Please write back if you get any problems with the configuration.

**Reply by marques-work** [reply]
Hi @mrexox, so sorry this reply is very late. I created a [new repo](https://github.com/marques-work/lefthook-interactive-test) with just these files, and I still can't get it to work. I installed `lefthook` 1.6.11 using `go`'s package installer.

I see the same behavior: it prints the prompt and just hangs indefinitely.

![image](https://github.com/evilmartians/lefthook/assets/723642/729d7497-d16d-49d8-936c-b26e737b4bdf)

It doesn't seem to take any input (typing isn't echoed to the screen, and hitting `enter` afterward has no effect). I need to hit `ctrl-C` to abort when this happens.

I wonder if it's either platform specific or if there's something weird with my environment?

I'm on linux (PopOS), running bash `5.1.16(1)-release`.

Noteworthy: if I run the script directly in my terminal (not through lefthook), it works as expected.

**Reply by marques-work** [reply]
Update 2: I created a completely new user on my system with a vanilla `.bashrc`. It should be running with all defaults -- no customization, completely clean.

Installed golang 1.22.3 and used the same installation method (go's package manager) to install lefthook. Cloned the repo I linked in my previous comment and tried `git commit --allow-empty` as a test.

Same result -- lefthook just hangs and doesn't accept any input.

**Reply by mrexox** [reply]
Thank you for providing a repo example. I think I could find the issue and I will try to prepare a fix ASAP.

**Reply by mrexox** [reply, marked as answer]
Please, check the new 1.6.12 version of lefthook.

**Reply by marques-work** [reply]
Yes, it works! Thanks so much for the quick fix!

**Comment by marques-work** [comment]
Update 3: tried with a fresh EC2 (`Amazon Linux release 2023.4.20240513`) and performed the same test.

Same result (hangs forever):

![image](https://github.com/evilmartians/lefthook/assets/723642/2de8daeb-2806-460d-b3a7-6e549eaa54df)

EC2 is using bash `5.2.15(1)-release`.
